File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.java
class Solution {
    public int findMin(int[] nums) {
        int n = nums.length;
        int st = 0, end = n-1;
        int ans = -1;

        while(st <= end) {
            int mid = st + (end - st)/2;

            if(nums[mid] <= nums[n-1]) {
                ans = nums[mid];
                end = mid - 1;
            } else {
                st = mid + 1;
            }
        }
        return ans;
    }
}

File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
        int lo = 0,hi=nums.size()-1;
        while(lo<hi){
            int mid = lo +(hi-lo)/2;
            if(nums[mid]>nums[hi]){
                lo = mid+1;
            }else{
                hi=mid;  // may be possible that it may be the smallest element so we do hi=mid , insted of hi=mid-1
            }
        }
        return nums[lo];  // also return nums[hi]
    }
};